File: Έγγραφα SEM4105/Λύσεις ασκήσεων/07. Έβδομο εργαστήριο/1b/newton.cpp
#include <iostream>
#include <cmath>
#include <fstream>
#include <vector>
#include <algorithm>
#include <string>


using vec = std::vector<double>;

void create(std::string fname)
{
    std::size_t constexpr n{15};
    double constexpr a{2.0}, b{4.0}, step{(b-a)/(n-1)};
    
    std::ofstream out{fname};

    out << n << '\n';
    for (std::size_t i{0}; i < n; ++i) {
	double x{a + step * i};
	double y{std::sin(x)};
	out << x << ' ' << y << '\n';
    }
}



//     Read data file
void read_data(std::string fname, vec & x, vec & y)
{
    std::ifstream in{fname};

    std::size_t n;
    in >> n;
    
    x.reserve(n);
    y.reserve(n);

    for (std::size_t i{0}; i < n; ++i) {
	double a, b;
	in >> a >> b;
	x.push_back(a);
	y.push_back(b);
    }
}


double q(std::size_t i, vec const & x, double z)
{
    double r{1.0};
    for (std::size_t j{0}; j < i; ++j) {
	r *= z-x[j];
    }

    return r;
}


void coefficients(vec const & x, vec const & y, vec & a)
{
    for (std::size_t i{0}; i < y.size(); ++i) {
	double s{y[i]};
	for (std::size_t j{0}; j < i; ++j) {
	    s -= a[j] * q(j,x,x[i]);
	}
	a[i] = s / q(i, x, x[i]);
    }
}


double poly(vec const & a, vec const & x, double z)
{
    double r{0.0};
    for (std::size_t i{0}; i < a.size(); ++i) {
	r += a[i] * q(i,x,z);
    }
    return r;
}


int main()
{
    create("points.dat");
    
    vec x, y;
    read_data("points.dat", x,y);

    std::size_t constexpr m{100};

    auto const p = std::minmax_element(x.cbegin(), x.cend());

    double const & min = *(p.first);
    double const & max = *(p.second);
    double const step = (max-min)/(m-1);

    vec a(x.size());
    coefficients(x,y,a);
    
    for (std::size_t i{0}; i < m; ++i) {
	double xout = min + step * i;
	double yout = poly(a,x, xout);

	std::cout << xout << ' ' << yout << '\n';
    }
}

File: Έγγραφα SEM4105/Λύσεις ασκήσεων/07. Έβδομο εργαστήριο/1b/newton.f
      SUBROUTINE CREATE(FNAME)
      CHARACTER(*) FNAME

      INTEGER N
      PARAMETER (N = 15)
      DOUBLE PRECISION A, B, STEP
      PARAMETER(A = 2.0D0, B = 4.0D0, STEP = (B-A)/(N-1))
      DOUBLE PRECISION X, Y
      INTEGER I

      OPEN(22, FILE=FNAME)

      WRITE (22, *) N
      DO 10 I= 1, N
        X = A + STEP * (I-1)
        Y = SIN(X)

        WRITE(22,*) X, Y
 10   CONTINUE

      CLOSE(22)
      END

      SUBROUTINE READDA(FNAME, X, Y, N)
      CHARACTER(*) FNAME
      DOUBLE PRECISION X(0:*), Y(0:*)
      INTEGER N

      INTEGER I
      
      OPEN(22, FILE=FNAME)
      
      READ (22, *) N

      N = N-1
      
      DO 10 I = 0, N
        READ (22,*) X(I), Y(I)
 10   CONTINUE
      
      CLOSE(22)
      END 

      FUNCTION Q(I,X,Z)
      INTEGER I
      DOUBLE PRECISION X(0:*), Z, Q

      INTEGER J
      
      Q = 1D0
      DO 10 J = 0, I-1
        Q = Q * (Z-X(J))
 10   CONTINUE

      END
      
      
      SUBROUTINE COEFS(N, X, Y, A)
      INTEGER N      
      DOUBLE PRECISION X(0:*), Y(0:*), A(0:*)

      DOUBLE PRECISION S, Q
      INTEGER I
      EXTERNAL Q
      
      DO 20 I = 0, N        
        S = 0D0
        DO 10 J = 0, I-1
          S = S + A(J) * Q(J, X, X(I)) 
 10     CONTINUE
        
        A(I) = (Y(I) - S) / Q(I, X, X(I))
 20   CONTINUE
      
      END 


      FUNCTION POLY(N, A, X, Z)
      INTEGER N
      DOUBLE PRECISION A(0:*), X(0:*), Z
      DOUBLE PRECISION POLY

      DOUBLE PRECISION Q
      EXTERNAL Q

      INTEGER I

      POLY = 0D0
      DO 10 I = 0, N
        POLY = POLY + A(I) * Q(I, X, Z)
 10   CONTINUE
      
      END





      FUNCTION MAXVAL(N,X)
      INTEGER N
      DOUBLE PRECISION X(0:*), MAXVAL

      DOUBLE PRECISION P
      INTEGER I

      P=X(0)
      DO 10 I=1,N
        IF (P .LT. X(I)) P = X(I)
 10   CONTINUE
      
      MAXVAL = P
      END

      FUNCTION MINVAL(N,X)
      INTEGER N
      DOUBLE PRECISION X(0:*), MINVAL

      DOUBLE PRECISION P
      INTEGER I

      P=X(0)
      DO 10 I=1,N
        IF (P .GT. X(I)) P = X(I)
 10   CONTINUE

      MINVAL = P
      END



      PROGRAM NEWTON
      DOUBLE PRECISION X(1000), Y(1000), A(1000)
      INTEGER M
      PARAMETER(M = 100)
      DOUBLE PRECISION MINX, MAXX, STEP, XOUT, YOUT
      INTEGER I, N
      DOUBLE PRECISION POLY, MAXVAL, MINVAL
      EXTERNAL MAXVAL, MINVAL, CREATE, READDA, POLY

      CALL CREATE("points.dat")
      CALL READDA("points.dat", X, Y, N)
      
      CALL COEFS(N,X,Y,A)


      MINX = MINVAL(N,X)
      MAXX = MAXVAL(N,X)
      STEP = (MAXX - MINX) / (M-1)

      DO 10 I = 1,M
        XOUT = MINX + STEP * (I-1)
        YOUT = POLY(N, A, X, XOUT)

        PRINT *, XOUT, YOUT
 10   CONTINUE

      END 

